k2/csrc/fsa.cu

```diff
@@ -177,33 +179,104 @@ int32_t GetFsaVecBasicProperties(FsaVec &fsa_vec) {
   return properties;
 }
 
-Fsa FsaFromTensor(Tensor &t, bool *error) {
-  *error = false;
-  if (t.GetDtype() != kInt32Dtype) {
-    K2_LOG(WARNING) << "Could not convert tensor to FSA, wrong dtype, got "
-                    << TraitsOf(t.GetDtype()).Name() << " but expected "
-                    << TraitsOf(kInt32Dtype).Name();
+FsaVec FsaVecFromFsa(const Fsa &fsa) {
+  ContextPtr c = fsa.values.Context();
+  K2_CHECK(fsa.NumAxes() == 2);
+  RaggedShape first_axis = TrivialShape(c, fsa.shape.Dim0());
+  RaggedShape fsa_vec_shape = ComposeRaggedShapes(first_axis, fsa.shape);
+  return Ragged<Arc>(fsa_vec_shape, fsa.values);
+}
+
+int32_t GetFsaBasicProperties(const Fsa &fsa) {
+  if (fsa.NumAxes() != 2)
+    return 0;
+  FsaVec vec = FsaVecFromFsa(fsa);
+  return GetFsaVecBasicProperties(vec);
+}
+
+
+Fsa FsaFromArray1(Array1<Arc> &array, bool *error) {
+  const Arc *arcs_data = reinterpret_cast<const Arc *>(array.Data());
+  ContextPtr c = array.Context();
+  const int32_t num_arcs = array.Dim();
+
+
+  // If the FSA has arcs entering the final state, that will
+  // tell us what the final-state id is.
+  // If there are no arcs entering the final-state, we let the final state be
```

Comment:
The comments are actually incorrect, i.e., do not match the implementation.
The implementation assumes that there is at least one arc entering
the final state.

Reply:
You are right, I'll try to fix this.